File: 0110-balanced-binary-tree/0110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isBalanced(TreeNode root) {
        if( root ==null) return true;
        return depth(root)!=-1;
    }
    public int depth(TreeNode root){
        if(root ==null) return 0;
        int leftSize = depth(root.left);
        int rightSize =depth(root.right);
        if(leftSize ==-1 || rightSize == -1 || Math.abs(leftSize-rightSize)>1)
            return -1;
        return 1 + Math.max(leftSize,rightSize);
    }
}

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        int max_d=0;
        if(root==NULL) return true;
        return diameter(root,max_d)!=-1;
    }
    int diameter(TreeNode* root,int &max_d){
        if(root==NULL) return 0;
        int leftD=diameter(root->left,max_d);
        int rightD=diameter(root->right,max_d);
        if(leftD ==-1 || rightD == -1 || abs(leftD-rightD)>1) return -1;
        return 1+max(leftD,rightD);
    }
};